scripts/node_exporter_download.py

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 ScyllaDB
+#
+
+#
+# This file is part of Scylla.
+#
+# Scylla is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Scylla is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import tarfile
+import io
+import requests
+import shutil
+from pathlib import Path
+
+VERSION='1.0.1'
+DEST_DIR='{}/../build/node_exporter'.format(os.path.dirname(__file__))
+
+if __name__ == '__main__':
+    if os.path.exists(DEST_DIR):
+        shutil.rmtree(DEST_DIR)
+    os.makedirs(DEST_DIR)
+    r = requests.get(f'https://github.com/prometheus/node_exporter/releases/download/v{VERSION}/node_exporter-{VERSION}.linux-amd64.tar.gz', timeout=5)
+    bo = io.BytesIO(r.content)
```

Comment:
I don't know that I want a rule that uses the network in our build system. Yes, we have them in jmx and java, but let's not add more.

Reply:
Hmm, I thought building from source can be the solution for this, but I found that node_exporter and prometheus build system are actually using network to fetch go dependencies (like our java project)
We may have to add prebuilt binary to this repo if we really want to make it completely offline.

Reply:
Or we may require the user to fetch and install node_exporter themselves.

Reply:
How is download node_exporter on dbuild docker image building time, instead of source code building time?
As we already do `pip3 install cassandra-driver` for python module, we also can `go get  github.com/prometheus/node_exporter@v1.0.1` to install node_exporter on dbuild Docker image (stored on ~/go).
Once it installed, we can just copy it on building time, no need network.

Additionally, we may able to do same thing for Java, by resolving all maven dependencies (stored on ~/.m2).
referenced: https://stackoverflow.com/questions/42208442/maven-docker-cache-dependencies

Reply:
How about following change to add node_exporter on dbuild image:
```
diff --git a/install-dependencies.sh b/install-dependencies.sh
index 62c84c821..7ff717235 100755
--- a/install-dependencies.sh
+++ b/install-dependencies.sh
@@ -91,6 +91,7 @@ fedora_packages=(
     fakeroot
     file
     dpkg-dev
+    golang
 )
 
 fedora_python3_packages=(
@@ -194,6 +195,7 @@ elif [ "$ID" = "fedora" ]; then
     fi
     yum install -y "${fedora_packages[@]}" "${fedora_python3_packages[@]}"
     pip3 install cassandra-driver
+    GO111MODULE=on go get github.com/prometheus/node_exporter@v1.0.1
 elif [ "$ID" = "centos" ]; then
     yum install -y "${centos_packages[@]}"
     echo -e "Configure example:\n\tpython3.4 ./configure.py --enable-dpdk --mode=release --static-boost --compiler=/opt/scylladb/bin/g++-7.3 --python python3.4 --ldflag=-Wl,-rpath=/opt/scylladb/lib64 --cflags=-I/opt/scylladb/include --with-antlr3=/opt/scylladb/bin/antlr3"
diff --git a/tools/toolchain/Dockerfile b/tools/toolchain/Dockerfile
index 9f5782d31..7ac1a8ff3 100644
--- a/tools/toolchain/Dockerfile
+++ b/tools/toolchain/Dockerfile
@@ -6,10 +6,11 @@ RUN dnf -y install 'dnf-command(copr)' \
     && dnf -y install ccache \
     && dnf -y install gdb \
     && dnf -y install devscripts debhelper fakeroot file rpm-build \
-    && /bin/bash -x ./install-dependencies.sh && dnf -y update && dnf clean all \
+    && env GOPATH=/go /bin/bash -x ./install-dependencies.sh && dnf -y update && dnf clean all \
     && echo 'ALL ALL=(ALL:ALL) NOPASSWD: ALL' >> /etc/sudoers \
     && cp system-auth /etc/pam.d \
     && echo 'Defaults !requiretty' >> /etc/sudoers
 RUN mkdir -p /root/.m2/repository
 ENV JAVA8_HOME=/usr/lib/jvm/java-1.8.0-openjdk
+ENV GOPATH=/go
 CMD /bin/bash
```

Reply:
It means the build won't work outside dbuild (unless you run the same command).

We could move it to install-dependencies.sh (and then we can just download the package in install-dependencies.sh and store it in /opt/scylladb/dependencies or something).

Reply:
I guess that's similar to what we do with python-cassandra-driver.

Reply:
> It means the build won't work outside dbuild (unless you run the same command).

Right, and I'm trying to avoid breaking build outside of dbuild, something like this:
`go get` will download the package on `$GOPATH` if it's defined, otherwise it will download to `$HOME/go`.
So we can define both pattern in create-relocatable-package.py:
```
if os.getenv('GOPATH'):
    node_exporter = '{}/bin/node_exporter'.format(os.getenv('GOPATH'))
else:
    node_exporter = os.path.expanduser('~/go/bin/node_exporter')
ar.reloc_add(node_exporter, arcname='node_exporter')
```

On this way, we can support all these three patterns:
 - In dbuild, GOPATH=/go
 - Outside of dbuild, GOPATH defined
 - Outside of dbuild, GOPATH not defined, installed to ~/go

Note that why I defined `ENV GOPATH=/go` in Dockerfile is, because install-dependencies.sh will run in root, but building package could run in non-root user, so root and non-root have to reference same directory.

Reply:
I don't want separate paths for dbuild and outside dbuild. But I think we can just curl the binary package and store it somewhere, we don't need to build from source.

Same as install-dependencies.sh doesn't build clang from source, it just gets a binary package. If one day we get node_exporter in Fedora, we can switch it to rpm.

Reply:
Oh sorry, I was misunderstanding that `go get` just downloading source&binary, if it's compiling we should use just download binary tar.gz using curl instead.
How about something like this:
```
diff --git a/install-dependencies.sh b/install-dependencies.sh
index 62c84c821..4573dbe7b 100755
--- a/install-dependencies.sh
+++ b/install-dependencies.sh
@@ -91,6 +91,7 @@ fedora_packages=(
     fakeroot
     file
     dpkg-dev
+    curl
 )
 
 fedora_python3_packages=(
@@ -140,6 +141,8 @@ arch_packages=(
     thrift
 )
 
+NODE_EXPORTER_VERSION=1.0.1
+
 print_usage() {
     echo "Usage: install-dependencies.sh [OPTION]..."
     echo ""
@@ -194,6 +197,10 @@ elif [ "$ID" = "fedora" ]; then
     fi
     yum install -y "${fedora_packages[@]}" "${fedora_python3_packages[@]}"
     pip3 install cassandra-driver
+    curl -fSL -o build/node_exporter.tar.gz https://github.com/prometheus/node_exporter/releases/download/v$NODE_EXPORTER_VERSION/node_exporter-$NODE_EXPORTER_VERSION.linux-amd64.tar.gz
+    tar -C build -xpf build/node_exporter.tar.gz
+    rm -rf build/node_exporter
+    mv build/node_exporter-$NODE_EXPORTER_VERSION.linux-amd64 build/node_exporter
 elif [ "$ID" = "centos" ]; then
     yum install -y "${centos_packages[@]}"
     echo -e "Configure example:\n\tpython3.4 ./configure.py --enable-dpdk --mode=release --static-boost --compiler=/opt/scylladb/bin/g++-7.3 --python python3.4 --ldflag=-Wl,-rpath=/opt/scylladb/lib64 --cflags=-I/opt/scylladb/include --with-antlr3=/opt/scylladb/bin/antlr3"
```

Reply:
We can't use build/ in install-dependencies.sh, it doesn't exist yet. . Also, we don't know which directory the user will use after ./install-dependencies is run (I have two scylla directories on my machine). We can store it in /opt/scylladb/dependencies.